### PR TITLE
Update JWT recommendations

### PR DESCRIPTION
**Pull request by jmandel**
This addresses #43, #26, #18.

### CODE REVIEW

File: authorization.md [hunk 1]
```diff
@@ -173,14 +227,26 @@ The access token response is a JSON object, with the following properties:
...
+* check that the JWT `exp` claim is valid
+* check that the JWT `aud` claim matches the server's OAuth token URL (the URL to which the token was `POST`ed)
 * check that this is not a `jti` value previously encountered for the given `sub` within the maximum allowed authentication JWT lifetime (5 minutes). This check prevents replay attacks.
-* ensure that the `client_id` provided is known and associated with the supplied `iss`
+* ensure that the `client_id` provided is known matches the JWT's `iss` claim
+
+To resolve a key to verify signatures, a server follows this algorithm:
+1. If `jku` header is present, verify that the `jku` is whitelisted (i.e., that it 
+matches the value supplied at registration time for the specified `client_id`).
+a. If the `jku` header  is whitelisted, the signature verification fails.
```
Comment (jeffy-grh) [inline authorization.md]: Meant `If the ``jku`` header is not whitelisted, the signature verification fails.`?

Comment (jmandel) [reply authorization.md]: Yes -- thanks! Fixing :-)

**Comment by jmandel** [comment]
Updated to address https://github.com/smart-on-fhir/fhir-bulk-data-docs/issues/17 as well.

File: authorization.md [hunk 2]
```diff
@@ -171,16 +226,37 @@ The access token response is a JSON object, with the following properties:
...
-Servers SHALL
+In addition, we require that servers SHALL:
 * validate the signature on the JWT
-* check that the JWT `exp` is valid
-* check that the JWT `aud` matches the server's OAuth token URL (the URL to which the token was `POST`ed)
+* check that the JWT `exp` claim is valid
+* check that the JWT `aud` claim matches the server's OAuth token URL (the URL to which the token was `POST`ed)
 * check that this is not a `jti` value previously encountered for the given `sub` within the maximum allowed authentication JWT lifetime (5 minutes). This check prevents replay attacks.
-* ensure that the `client_id` provided is known and associated with the supplied `iss`
+* ensure that the `client_id` provided is known and matches the JWT's `iss` claim
```
Comment (gotdan) [inline authorization.md]: Shouldn't the client_id be in the `sub` claim?

File: authorization.md [hunk 3]
```diff
@@ -39,14 +40,35 @@ consider the [OAuth 2.0 Dynamic Client Registration
...
+while this specification recommends algorithms for interoperability, it does
+not mandate any algorithm.
 
-Upon registration, the server assigns a `client_id`, which 
-the client uses when obtaining an access token.
+No matter how a JWKS is communicated to the EHR, each key in the JWKS *must be* an asymmetric
+key whose content is conveyed using "bare key" properties (i.e., direct base64 encoding of 
+key material as integer values). This means that:
+
+* For RSA public keys, each MUST include `n` and `e` values (modulus and exponent)
```
Comment (gotdan) [inline authorization.md:66]: Should we update the worked example below to match this - it's 404ing on the public key, but in any case the one included in the original backend services docs on the SMART site isn't a JWKS?

File: authorization.md [hunk 4]
```diff
@@ -171,16 +226,37 @@ The access token response is a JSON object, with the following properties:
...
+Servers SHALL follow all requirements defined in [Section 3 of RFC7523](https://tools.ietf.org/html/rfc7523#section-3).
 
-Servers SHALL
+In addition, we require that servers SHALL:
 * validate the signature on the JWT
-* check that the JWT `exp` is valid
-* check that the JWT `aud` matches the server's OAuth token URL (the URL to which the token was `POST`ed)
+* check that the JWT `exp` claim is valid
+* check that the JWT `aud` claim matches the server's OAuth token URL (the URL to which the token was `POST`ed)
 * check that this is not a `jti` value previously encountered for the given `sub` within the maximum allowed authentication JWT lifetime (5 minutes). This check prevents replay attacks.
```
Comment (lcmaas) [inline authorization.md]: the jti check should be for the given `iss` (not `sub`) even if iss and sub happen to be the same.

File: authorization.md (hunk 2 above)
Comment (jmandel) [reply authorization.md]: Our current advice is that they are both the same.

File: authorization.md (hunk 4 above)
Comment (jmandel) [reply authorization.md]: Agreed - will fix. Thanks!

File: authorization.md (hunk 3 above)
Comment (jmandel) [reply authorization.md:66]: Yes, we should update examples. I will do this before merging.

File: authorization.md (hunk 2 above)
Comment (gotdan) [reply authorization.md]: Hmm, the table above says:

iss | required | The service's issuer URI, as registered with the EHR's authorization server
sub | required | The service's client_id, as determined during registration with the EHR's authorization server

Do we need to clarify that the `iss` is the client_id there?

Comment (jmandel) [reply authorization.md]: Clarifying now :-)

File: authorization.md (hunk 3 above)
Comment (jmandel) [reply authorization.md:66]: Update the example and added `authorization-example-jwks-and-signatures.ipynb` and `authorization-example-jwks.json`.

**Pull request merged**